packages/cw721/src/state.rs

```diff
@@ -0,0 +1,460 @@
+        let val = self.token_count(storage)? - 1;
+        self.token_count.save(storage, &val)?;
+        Ok(val)
+    }
+}
+
+pub fn token_owner_idx<TNftMetadataExtension>(
+    _pk: &[u8],
+    d: &NftInfo<TNftMetadataExtension>,
+) -> Addr {
+    d.owner.clone()
+}
+
+#[cw_serde]
+pub struct NftInfo<TNftMetadataExtension> {
+    /// The owner of the newly minted NFT
+    pub owner: Addr,
+    /// Approvals are stored here, as we clear them all upon transfer and cannot accumulate much
+    pub approvals: Vec<Approval>,
+
+    /// Universal resource identifier for this NFT
+    /// Should point to a JSON file that conforms to the ERC721
+    /// Metadata JSON Schema
+    pub token_uri: Option<String>,
+
+    /// You can add any custom metadata here when you extend cw721-base
+    pub extension: TNftMetadataExtension,
+}
+
+#[cw_serde]
+pub struct Approval {
+    /// Account that can transfer/send the token
+    pub spender: Addr,
+    /// When the Approval expires (maybe Expiration::never)
+    pub expires: Expiration,
+}
+
+impl Approval {
+    pub fn is_expired(&self, block: &BlockInfo) -> bool {
+        self.expires.is_expired(block)
+    }
+}
+
+pub struct TokenIndexes<'a, TNftMetadataExtension>
+where
+    TNftMetadataExtension: Cw721State,
+{
+    pub owner: MultiIndex<'a, Addr, NftInfo<TNftMetadataExtension>, String>,
+}
+
+impl<'a, TNftMetadataExtension> IndexList<NftInfo<TNftMetadataExtension>>
+    for TokenIndexes<'a, TNftMetadataExtension>
+where
+    TNftMetadataExtension: Cw721State,
+{
+    fn get_indexes(
+        &'_ self,
+    ) -> Box<dyn Iterator<Item = &'_ dyn Index<NftInfo<TNftMetadataExtension>>> + '_> {
+        let v: Vec<&dyn Index<NftInfo<TNftMetadataExtension>>> = vec![&self.owner];
+        Box::new(v.into_iter())
+    }
+}
+
+#[cw_serde]
+pub struct CollectionMetadata<TCollectionMetadataExtension> {
+    pub name: String,
+    pub symbol: String,
+    pub extension: TCollectionMetadataExtension,
+    pub updated_at: Timestamp,
+}
+
+impl<TCollectionMetadataExtension, TCollectionMetadataExtensionMsg>
+    StateFactory<CollectionMetadata<TCollectionMetadataExtension>>
+    for CollectionMetadataMsg<TCollectionMetadataExtensionMsg>
+where
+    TCollectionMetadataExtension: Cw721State,
+    TCollectionMetadataExtensionMsg: Cw721CustomMsg + StateFactory<TCollectionMetadataExtension>,
+{
+    fn create(
+        &self,
+        deps: Deps,
+        env: &cosmwasm_std::Env,
+        info: &cosmwasm_std::MessageInfo,
+        current: Option<&CollectionMetadata<TCollectionMetadataExtension>>,
+    ) -> Result<CollectionMetadata<TCollectionMetadataExtension>, Cw721ContractError> {
+        self.validate(deps, env, info, current)?;
+        match current {
+            // Some: update existing metadata
+            Some(current) => {
+                let mut updated = current.clone();
+                if let Some(name) = &self.name {
+                    updated.name = name.clone();
+                }
```

Comment:
... then in `create` of StateFactory, it:

- validates msg first,
- uses current state and
- update it with msg data

... then it goes one level deeper, ...